From: Filipe Manana <fdmanana@suse.com>
Date: Wed, 27 Jul 2016 16:25:13 +0100
Patch-mainline: Not yet, used for debugging issues in openqa only
References: bsc#990384
Subject: [PATCH] Btrfs: add warning on failure to do delayed update of inodes

If we fail to do a delayed update of an inode (either due to a transient
-ENOMEM or -ENOSPC) we can end up allowing for persisting modified,
deleted or new extents/xattrs/links without persisting an updated inode
item. This allows for rare and short time windows where inconsistent
snapshots are created that when used for send operations trigger a BUG_ON.
So add a limited rate warning when this happens with the purpose of
validating if this is what's happening at bsc#990384.

Signed-off-by: Filipe Manana <fdmanana@suse.com>
---
 fs/btrfs/inode.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/fs/btrfs/inode.c b/fs/btrfs/inode.c
index 7210d58..acab92b 100644
--- a/fs/btrfs/inode.c
+++ b/fs/btrfs/inode.c
@@ -3909,6 +3909,10 @@ noinline int btrfs_update_inode(struct btrfs_trans_handle *trans,
 		ret = btrfs_delayed_update_inode(trans, root, inode);
 		if (!ret)
 			btrfs_set_inode_last_trans(trans, inode);
+		else
+			btrfs_warn_rl(root->fs_info,
+				      "failed to do delayed inode update for inode %llu root %llu, error %d",
+				      btrfs_ino(inode), root->root_key.objectid, ret);
 		return ret;
 	}
 
-- 
1.8.4.5

